video/buffer_stream.h

```diff
@@ -0,0 +1,102 @@
+#ifndef BUFFER_STREAM_H
+#define BUFFER_STREAM_H
+
+#include <memory>
+#include <Stream.h>
+
+#include "types.h"
+
+class BufferStream : Stream {
+	public:
+		BufferStream(uint32_t bufferLength);
+		int available();
+		int read();
+		int peek();
+		size_t write(uint8_t b);
+
+		void rewind();
+
+		inline uint8_t * getBuffer() {
+			return buffer.get();
+		}
+		inline uint32_t size() {
+			return bufferLength;
+		}
+		bool writeBuffer(uint8_t * data, uint32_t length, uint32_t offset);
+		void writeBufferByte(uint8_t data, uint32_t offset);
+		bool incrementBufferByte(uint32_t offset, int8_t by);
+	private:
+		std::unique_ptr<uint8_t[]> buffer;
+		uint32_t bufferLength;
+		uint32_t bufferPosition;
+};
+
+BufferStream::BufferStream(uint32_t bufferLength) : bufferLength(bufferLength) {
+	buffer = make_unique_psram_array<uint8_t>(bufferLength);
+	rewind();
+}
+
+int BufferStream::available() {
+	return bufferLength - bufferPosition;
+}
+
+int BufferStream::read() {
+	if (bufferPosition < bufferLength) {
+		return buffer[bufferPosition++];
+	}
+	return -1;
+}
+
+int BufferStream::peek() {
+	if (bufferPosition < bufferLength) {
+		return buffer[bufferPosition];
+	}
+	return -1;
+}
+
+void BufferStream::rewind() {
+	bufferPosition = 0;
+}
+
+size_t BufferStream::write(uint8_t b) {
+	// write is not supported
+	return 0;
+}
```

Comment:
not supporting write on either BufferStream (or MultiBufferStream) is essentially a decision based on short-term practicality.  it does however mean that none of the VDU calls that write back to the z80 will actually successfully write when they're being processed via a buffer.

for the most part I believe that should be harmless.  longer term tho, buffers could be created with the z80 comms channel passed to them so they _can_ write back

Reply:
this subject has had considerably more thought now since this comment was written

for VDU stream processors, input and output streams are now separated, and a command has been added to set the output stream.  the top-level VDU stream processor will set its output stream by default to be the eZ80 comms channel.  output can however be sent to a buffer instead.

BufferStream does not support `write`, and nor will it.  a new WritableBufferStream has been added, which can be created with a new "create buffer" command.  this split is intended to help prevent overwriting of command buffers with data.

when a command buffer is called, a new VDU stream processor will be created with that buffer as the input stream, and the output stream set as the _current_ output stream.  usually that will mean the output stream remains the eZ80 comms cannel.

setting the output to a (writable) buffer allows for the capture of data packets that would otherwise have been sent to the eZ80.  combining this with other features coming soon to copy values between buffers, and conditionally execute buffers, will allow for some reasonably sophisticated behaviour